src/c2s/mongoose_c2s_ranch.erl

```diff
@@ -0,0 +1,137 @@
+            {ok, State#state{transport = TlsMod, socket = TlsSocket}};
+        {error, Reason} ->
+            {error, Reason}
+    end.
+
+tcp_to_tls(fast_tls, TcpSocket, TlsOpts) ->
+    case fast_tls:tcp_to_tls(TcpSocket, TlsOpts) of
+        {ok, TlsSocket} ->
+            fast_tls:recv_data(TlsSocket, <<>>),
+            {ok, TlsSocket};
+        Other -> Other
+    end;
+tcp_to_tls(just_tls, TcpSocket, TlsOpts) ->
+    case ranch_ssl:handshake(TcpSocket, TlsOpts, 1000) of
+        {ok, TlsSocket, _} -> {ok, TlsSocket};
+        Other -> Other
+    end.
+
+-spec socket_handle_data(state(), {tcp | ssl, term(), iodata()}) ->
+  iodata() | {raw, [exml:element()]} | {error, term()}.
+socket_handle_data(#state{transport = fast_tls, socket = TlsSocket}, {tcp, _Socket, Data}) ->
+    mongoose_metrics:update(global, [data, xmpp, received, encrypted_size], iolist_size(Data)),
+    case fast_tls:recv_data(TlsSocket, Data) of
+        {ok, DecryptedData} ->
+            DecryptedData;
+        {error, Reason} ->
+            {error, Reason}
+    end;
+socket_handle_data(#state{transport = ranch_ssl, socket = Socket}, {ssl, Socket, Data}) ->
+    mongoose_metrics:update(global, [data, xmpp, received, encrypted_size], iolist_size(Data)),
+    Data;
+socket_handle_data(#state{transport = ranch_tcp, socket = Socket}, {tcp, Socket, Data}) ->
+    Data.
+
+-spec socket_activate(state()) -> ok.
+socket_activate(#state{transport = fast_tls, socket = Socket}) ->
+    fast_tls:setopts(Socket, [{active, once}]);
+socket_activate(#state{transport = ranch_ssl, socket = Socket}) ->
+    ranch_ssl:setopts(Socket, [{active, once}]);
+socket_activate(#state{transport = ranch_tcp, socket = Socket}) ->
+    ranch_tcp:setopts(Socket, [{active, once}]).
+
+-spec socket_close(state()) -> ok.
+socket_close(#state{transport = fast_tls, socket = Socket}) ->
+    fast_tls:close(Socket);
+socket_close(#state{transport = ranch_ssl, socket = Socket}) ->
+    ranch_ssl:close(Socket);
+socket_close(#state{transport = ranch_tcp, socket = Socket}) ->
+    ranch_tcp:close(Socket).
+
+-spec socket_send_xml(state(), iodata() | exml_stream:element() | [exml_stream:element()]) ->
+    ok | {error, term()}.
+socket_send_xml(#state{transport = Transport, socket = Socket}, XML) ->
+    Text = exml:to_iolist(XML),
+    case Transport:send(Socket, Text) of
+        ok ->
+            mongoose_metrics:update(global, [data, xmpp, sent, xml_stanza_size], iolist_size(Text)),
+            ok;
+        Error ->
+            Error
+        end.
+
+-spec has_peer_cert(mongoose_c2s_socket:state(), mongoose_listener:options()) -> boolean().
+has_peer_cert(#state{transport = fast_tls, socket = Socket}, #{tls := TlsOpts}) ->
+    case {fast_tls:get_verify_result(Socket), fast_tls:get_peer_certificate(Socket), TlsOpts} of
+        {0, {ok, _}, _} -> true;
+        {18, {ok, _}, #{verify_mode := selfsigned_peer}} -> true;
+        {_, {ok, _}, _} -> false;
+        {_, error, _} -> false
```

Comment:
Here you could move the comment from `mongoose_c2s_socket` that says `%% 18 is OpenSSL's and fast_tls's error code for self-signed certs`, so that that `18` ain't so magic.